### ISSUE TITLE
Can not get the 'journald' plugin with correct name.

### ISSUE THREAD

**Issue by sweet9** [kubernetes/node-problem-detector#389]
Hi, recently I run the node problem detector in my local environment when I specified the go launch arguments like this:
`--logtostderr=true --v=5 --config.system-log-monitor config/kernel-monitor.json,config/docker-monitor.json --enable-k8s-exporter=false`
As you see, I wonder registry and launch two log watchers to match something, however, there is  still an error message when it is loading the second one (docker log watcher):
`No create function found for plugin "**journald**"`

the name of plugin which is in config/docker.monitor.json is 'Journald',
`{
	"plugin": "journald",
	"pluginConfig": {
		"source": "dockerd"
	},
	"logPath": "/var/log/journal",
	......`
all of them look like correct, BUT when the code as following are running:
`// The name here has been passed as 'filelog'
func registerLogWatcher(name string, create types.WatcherCreateFunc) {
	createFuncs[name] = create
}`
The method registerLogWatcher() will register a watcher with name '**filelog**',
It can explain why the method GetLogWatcherOrDie(config types.WatcherConfig) has an error message above. there is only a watcher named 'filelog' in the registry variable.

**Comment by xueweiz** [comment]
/assign @xueweiz 

The "journald" log watcher is registered [here](https://github.com/kubernetes/node-problem-detector/blob/f479d09e581e1dd294672c9c997c566c7a5dbf81/pkg/systemlogmonitor/logwatchers/register_journald.go#L29):
`registerLogWatcher(journaldPluginName, journald.NewJournaldWatcher)`

If you don't have the "journald" plugin available, very likely is because you are compiling NPD while disabling the [`journald` build tag](https://github.com/kubernetes/node-problem-detector/blob/f479d09e581e1dd294672c9c997c566c7a5dbf81/pkg/systemlogmonitor/logwatchers/register_journald.go#L1).

A few questions:
1. Can you show me how are you compiling NPD? i.e. some command like:
`BUILD_TAGS="disable_custom_plugin_monitor disable_system_stats_monitor" make` or `BUILD_TAGS="-journald" make`

2. Which operating system are you running NPD on? Can you run `cat /etc/os-release` and paste it here? (which will give me the more detailed OS versioning)

3. Does your OS has/uses [systemd-journald](https://wiki.archlinux.org/index.php/Systemd/Journal)? To verify this, you just need to run `sudo journalctl` and see whether it show you the logs.

Thanks! :)

**Comment by sweet9** [comment]
I found that I haven't disable the journald build tag.
and  I start NPD with the local IDE by running the main method directly.

for your questions:
1. I haven't start NPD as a binary file, I start it with my IDE (goland) so that I can debug easier.

2.

**Comment by xueweiz** [comment]
Ah that explains it :)
We typically build NPD by running `Make`. You can see that in Makefile:
1. [`ENABLE_JOURNALD` is set to `1` by default](https://github.com/kubernetes/node-problem-detector/blob/d000c1b060803cb8b3172fdcf9aa9956db942d7e/Makefile#L51)
1. [If](https://github.com/kubernetes/node-problem-detector/blob/d000c1b060803cb8b3172fdcf9aa9956db942d7e/Makefile#L64) `ENABLE_JOURNALD` is `1`, `journald` will be [appended](https://github.com/kubernetes/node-problem-detector/blob/d000c1b060803cb8b3172fdcf9aa9956db942d7e/Makefile#L66) to `BUILD_TAGS`.
1.[ When we compile the NPD binary](https://github.com/kubernetes/node-problem-detector/blob/d000c1b060803cb8b3172fdcf9aa9956db942d7e/Makefile#L101), [BUILD_TAGS is used at the `-tags` parameter](https://github.com/kubernetes/node-problem-detector/blob/d000c1b060803cb8b3172fdcf9aa9956db942d7e/Makefile#L103) which is then passed to Golang as [build tags](https://golang.org/pkg/go/build/).

So long story short, it's best to use `make` to compile NPD, because that is the supported and [tested](https://github.com/kubernetes/node-problem-detector/blob/d000c1b060803cb8b3172fdcf9aa9956db942d7e/.travis.yml#L22) way. Any other way could be broken by future changes, simply because they are not tested :)

And if you **have** to use IDE for now, [adding `journald` to your build tags](https://www.jetbrains.com/help/go/configuring-build-constraints-and-vendoring.html) could help you workaround it. :)

From your logs above, your system does have support for journald. You might need to install a library for it (`libsystemd-journal-dev` or `libsystemd`), as documented in the [README](https://github.com/kubernetes/node-problem-detector#build-image).

**Comment by sweet9** [comment]
Thanks for your reply, finally I understand why we specify the BUILD_TAG in MakeFile.
and I install the lib tools like this:
`apt-get install libsystemd-dev`

And add the custom_tag in my IDE just as you said, It works well, Great.

but a new problem is here when the following code has been executed:
```
unc getJournal(cfg types.WatcherConfig, startTime time.Time) (*sdjournal.Journal, error) {
	var journal *sdjournal.Journal
	var err error
	if cfg.LogPath == "" {
		journal, err = sdjournal.NewJournal()
		if err != nil {
			return nil, fmt.Errorf("failed to create journal client from default log path: %v", err)
		}
		glog.Info("unspecified log path so using systemd default")
	} else {
		// If the path doesn't exist, NewJournalFromDir will
		// create it instead of returning error. So check the
		// path existence ourselves.
		if _, err = os.Stat(cfg.LogPath); err != nil {
			return nil, fmt.Errorf("failed to stat the log path %q: %v", cfg.LogPath, err)
		}
		// Get journal client from the log path.
		journal, err = sdjournal.NewJournalFromDir(cfg.LogPath)
		if err != nil {
			return nil, fmt.Errorf("failed to create journal client from path %q: %v", cfg.LogPath, err)
		}
	}
```
Error message : 
`failed to stat the log path "/var/log/journal": stat /var/log/journal: no such file or directory`

I guess that I need to modify the path in config/docker-monitor.json. 

Thanks.

**Comment by xueweiz** [comment]
Sounds good :) Glad it worked for you.

Closing this issue.
/close

**Comment by sweet9** [comment]
But I have a new question above after I enable the 'journald' log Watcher.
The watcher will read the /var/log/journal file, but it can read successfully in my local environment, the error message like this:
`failed to stat the log path "/var/log/journal": stat /var/log/journal: no such file or directory` 

I doubt the log path configured in /config/docker-monitor.json is incorrect when it has been used in my local environment. should I need to change it? although I am not sure.

**Comment by xueweiz** [comment]
From your previous [comment](https://github.com/kubernetes/node-problem-detector/issues/389#issuecomment-558917790), apparently that your system is using systemd-journald.

Since the error says `stat /var/log/journal: no such file or directory`, my guess is that your systemd-journald is configured to store the journal logs in some other directories (perhaps `/run/log/journal`).
 
See [journald configuration](https://www.freedesktop.org/software/systemd/man/journald.conf.html) to understand how is journald configured and how is the log stored on the disk.

**Comment by sweet9** [comment]
Thanks for your help, I found the message with path '/run/log/journal' on my local environment, 
And I added some rules for my scenario as the customized rules. Great! it works well.
